subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/DefaultTransformedFileCacheTest.groovy

```diff
@@ -345,6 +348,68 @@ class DefaultTransformedFileCacheTest extends ConcurrentSpec {
         0 * transform._
     }
 
+    def "marks used entries accessed by touching their output directory"() {
+        given:
+        def outputDir = null
+        def inputFile = tmpDir.file("a")
+        def transform = Mock(BiFunction) {
+            apply(_, _) >> { File file, File dir ->
+                outputDir = dir
+                def parent = new TestFile(dir, "dir")
+                parent.mkdirs()
+                [new TestFile(parent, "file").touch()]
+            }
+        }
+        snapshotter.snapshotAll(_) >> snapshot(HashCode.fromInt(234))
+
+        when:
+        def result = cache.getResult(inputFile, HashCode.fromInt(123), transform)
+
+        then:
+        result*.name == ["file"]
+        def cachedFile = result[0]
+        def cachedDir = cachedFile.parentFile
+        cachedDir.name == "dir"
+        cachedDir.parentFile == outputDir
+
+        when:
+        new TestFile(outputDir).makeOlder()
+        cachedFile.makeOlder()
+        cachedDir.makeOlder()
+        def modificationDateOfCachedFile = cachedFile.lastModified()
+        def modificationDateOfCachedDir = cachedDir.lastModified()
+        // not all filesystems support millisecond precision, but all support seconds
+        def timestampBeforeAccessInSeconds = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis())
```

Comment:
⭕️ This relies on no more than a second passing between here and the last assertion I guess? Probably reasonable, but a safer approach would be passing a clock into the cache and letting it use the time from that clock instead of calling the static currentTimeMillis.